Use Sentry `flush` to send events immediately

mokagio opened the pull request:
This PR removes our custom logic to implement the "send all enqueue events and wait" functionality in favor of using the new Sentry's native `flush` method.

One big disadvantage of this approach is that we lose the precise callback management, because [`flush` is blocking for as long as it takes to send all events or for a given timeout to run out](https://github.com/getsentry/sentry-cocoa/issues/1013#issuecomment-1243634969).

I used the demo app to verify our custom implementation works and the events it sends reach Sentry. In the screenshot below, the top two events have been sent from this branch. Also notice the third and fourth events have "(no value)" for release and environment. That's a bug introduced at some point when migrating to version Sentry version 7. I haven't looked into it yet.

![image](https://user-images.githubusercontent.com/1218433/196594554-94de84bc-1185-4ae7-8155-7c85aae02127.png)

I'm of two minds on whether to go for using `flush` under the hood. I really dislike how `flush` doesn't provide a callback out of the box, and how it results in a less precise UI state handling on our end. At the same time, relying on Sentry's native code instead of implementing our own logic accessing their private API simplifies and de-risks the library.

I'd like other folks input on this, but I'm tempted to adopt this implementation on the basis that `logErrorAndWait`, `logErrorImmediately`, and `logErrorsImmediately` are used very sparingly by the library's clients. Namely, [in one occasion in Jetpack/WordPress](https://github.com/wordpress-mobile/WordPress-iOS/search?q=logErrorAndWait) and in two occasions in WooCommerce](https://github.com/woocommerce/woocommerce-ios/search?q=logFatalErrorAndExit).

mokagio committed:
Bump Sentry minimum requirement to 7.26.0

mokagio committed:
Use Sentry's native `flush` to force sending events

Notice that, because `flush` doesn't support callbacks but instead
blocks the main thread up to a given timeout interval, we had to remove
our more refined callback mechanism and UI state management logic.

mokagio committed:
Remove code made unused by using Sentry's `flush`

This code was all custom interaction with the Sentry SDK for the purpose
of extracting information to add to events sent to Sentry "manually" via
HTTP requests. Now that we use Sentry's native method for that, all the
information tracking is handled by the Sentry SDK.

mokagio committed:
Add a `userInfo` value when testing `logErrorImmediately` in demo app

mokagio commented on Sources/Remote Logging/Crash Logging/CrashLogging.swift:
```diff
@@ -23,6 +23,9 @@ public class CrashLogging {
     /// crash log events will only be sent in Release builds.
     public static let forceCrashLoggingKey = "force-crash-logging"
 
+    // TODO: This could be made a configurable instance property, with default value in the `init`
+    public static let eventsFlushingTimeout: TimeInterval = 15
```
Waiting to act upon this till we decided whether to go down the Sentry `flush` route.

crazytonyli replied:
Any plan for addressing this TODO now that this PR is no longer draft?

mokagio replied:
Good point. Done in 30a873ed4e02b54b77c8e87b6e3fa693da41d250

mokagio commented on Sources/UI/Crash Logging/CrashLoggingView.swift:
```diff
@@ -97,14 +97,11 @@ extension CrashLoggingView {
-                }
+            try crashLogging.logErrorImmediately(
+                error,
+                userInfo: ["custom-userInfo-key": "custom-userInfo-value"]
```
I realized we never "tested" sending custom information, so I added it. Here's the result:

<img width="1023" alt="image" src="https://user-images.githubusercontent.com/1218433/196597311-505798ef-2558-4845-9c8c-37911bc5da9a.png">

mokagio commented on Sources/UI/Crash Logging/CrashLoggingView.swift:
```diff
@@ -97,14 +97,11 @@ extension CrashLoggingView {
+                error,
+                userInfo: ["custom-userInfo-key": "custom-userInfo-value"]
+            ) {
+                sendErrorAndWaitStatus = .success
```
If we decide to adopt `flush`, it might be appropriate to update this `enum` to remove the error-tracking state, since there's no longer any error information with `flush`.

mokagio commented on Sources/Remote Logging/Crash Logging/CrashLogging.swift:
```diff
@@ -168,54 +171,24 @@ public extension CrashLogging {
 
     /// Sends an `Event` to Sentry and triggers a callback on completion
-    func logErrorImmediately(_ error: Error, userInfo: [String: Any]? = nil, level: SentryLevel = .error, callback: @escaping (Result<Bool, Error>) -> Void) throws {
+    func logErrorImmediately(_ error: Error, userInfo: [String: Any]? = nil, level: SentryLevel = .error, callback: @escaping () -> Void) throws {
```
Surprisingly, I'm not getting any errors or warnings from the compiler because of these methods still having `throws` despite there not being any genuine `try` call in the chain.

If we commit to `flush`, we should remove `throws` as well.

crazytonyli commented on Sources/Remote Logging/Crash Logging/CrashLogging.swift:
```diff
@@ -168,54 +171,24 @@ public extension CrashLogging {
-            let didSucceed = 200...299 ~= (urlResponse as! HTTPURLResponse).statusCode
-            callback(.success(didSucceed))
-        }.resume()
+        SentrySDK.flush(timeout: CrashLogging.eventsFlushingTimeout)
```
Putting this into a background queue so that we don't block `logErrorsImmediately` callers?

crazytonyli replied:
Or adding a `wait: Bool` argument, which'd make the `logErrorAndWait` simpler to implement.

mokagio committed:
Move call to Sentry's `flush` to a background queue

mokagio commented on Sources/Remote Logging/Crash Logging/CrashLogging.swift:
```diff
@@ -187,8 +187,12 @@ public extension CrashLogging {
+        queue.async {
+            SentrySDK.flush(timeout: CrashLogging.eventsFlushingTimeout)
+            callback()
+        }
```
@crazytonyli what do you think of this as an implementation of your suggestion to make the call in a background queue?

mokagio replied:
In regards to your suggestion to [add a `wait: Bool`](https://github.com/Automattic/Automattic-Tracks-iOS/pull/232#discussion_r999047433) parameter, are you thinking of something along the lines [of WordPress iOS](https://github.com/wordpress-mobile/WordPress-iOS/blob/feb1cc189cf70c050f8bc5aa74630d4b9dff34a2/WordPress/Classes/Utility/ContainerContextFactory.swift#L23-L35) where we use a `Bool` parameter to decide whether to call `context.performAndWait(_:)` or `context.perform(_:)`?

In this, I'm thinking if `wait == true` we wouldn't run the code in `queue.async`.

crazytonyli replied:
Yes, so that we don't need to use semaphore like the original implementation 😄

mokagio replied:
😕 The new diff makes this conversation odd because the code it referred to has moved. Anyways... @crazytonyli I had an initial go at adding a `wait` parameter. 

I did this in the form of a `private` func for the other methods to call.

```swift
private func logErrorsImmediately(
        _ errors: [Error],
        userInfo: [String: Any]? = nil,
        level: SentryLevel = .error,
        andWait wait: Bool,
        callback: @escaping () -> Void
    ) {
```

I didn't modify the original method signatures to avoid a breaking change, but also because I wasn't sure how it would have felt to have a `andWait:,` together with a `callback:`. If `wait == true`, then there's no need for a callback because the code that would run in there essentially runs once the method has finished anyway.

mokagio committed:
Refactor the various `logError...` implementation with a common one

This is based on the suggestion at
https://github.com/Automattic/Automattic-Tracks-iOS/pull/232#discussion_r1006282586

crazytonyli commented on Sources/Remote Logging/Crash Logging/CrashLogging.swift:
```diff
@@ -225,64 +186,40 @@ public extension CrashLogging {
+        if wait {
+            flushEventThenCallCallback()
+        } else {
+            let queue = DispatchQueue(label: "com.automattic.tracks.flushSentry", qos: .background)
```
Any reason why using a custom queue instead of `DispatchQueue.global`? Also, any reason why choose `.background` QoS, instead of the default one? (I almost always use the default QoS since I've heard scary things about QoS but don't really understand it 😸 )

mokagio replied:
I followed the suggestion from one of the Sentry engineers from [here](https://github.com/getsentry/sentry-cocoa/issues/1013#issuecomment-1244958158).

I trust your advice on this matter, though. I'll update.

mokagio replied:
Done in [Automattic/Automattic-Tracks-iOS@`8a0c84b` (#232)](https://github.com/Automattic/Automattic-Tracks-iOS/pull/232/commits/8a0c84b201c7b919aa6a2b8b752e0dcb3269c0a9)

crazytonyli commented on Sources/Remote Logging/Crash Logging/CrashLogging.swift:
```diff
@@ -168,54 +171,12 @@ public extension CrashLogging {
-            let didSucceed = 200...299 ~= (urlResponse as! HTTPURLResponse).statusCode
-            callback(.success(didSucceed))
-        }.resume()
+    func logErrorsImmediately(_ errors: [Error], userInfo: [String: Any]? = nil, level: SentryLevel = .error, callback: @escaping () -> Void) throws {
```
I had a quick look, it appears we can't get the result of `flush` call, is that correct? I think it would be good to at least have a "success or not" result, if we can't get a concrete error. But I guess there isn't anything we can do if Sentry doesn't expose such API.

mokagio replied:
Correct. See the details given [here](https://github.com/getsentry/sentry-cocoa/issues/1013). It's quite the let down. There must be a reason for the lack of callback in their [unified API](https://develop.sentry.dev/sdk/unified-api) rationale, but to be honest I haven't read that documentation.

mokagio committed:
Rename `eventsFlushTimeout` to `flushTimeout`

The word "events" made me think it ought to belong as part of
`EventLogging` which was confusing because performance monitoring is
currently implemented via Sentry (crash logging) instead.

mokagio committed:
Make `flushTimeout` a configurable instance property

mokagio committed:
Flush Sentry events on global `DispatchQueue` instead of a custom one

See discussion at:
https://github.com/Automattic/Automattic-Tracks-iOS/pull/232#discussion_r1009100622

mokagio committed:
Add `CHANGELOG.md` entry for the changes due to Sentry `flush`

mokagio committed:
Replace `success` and `error` cases with `done` in `SendErrorAndWaitStatus`

We no longer have error handling because of Sentry's `flush`.

mokagio committed:
Remove unused `throws` from `logError[s]Immediately`

None of the code left from the `flush` adoption threw errors.

mokagio committed:
Add `CHANGELOG.md` entry regarding the `throws` removal

The change in the `SendErrorAndWaitStatus` `enum` is not mentioned
because the type is `internal`.

mokagio committed:
Update `CrashLoggingView` for Sentry `flush` & remove `throws` (#236)
